tests/src/test/scala/com/nvidia/spark/rapids/AdaptiveQueryExecSuite.scala

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids
+
+import org.apache.spark.SparkConf
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.execution.{PartialReducerPartitionSpec, SparkPlan}
+import org.apache.spark.sql.execution.adaptive.{AdaptiveSparkPlanExec, AdaptiveSparkPlanHelper}
+import org.apache.spark.sql.execution.exchange.Exchange
+import org.apache.spark.sql.functions.when
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.rapids.execution.{GpuCustomShuffleReaderExec, GpuShuffledHashJoinBase}
+
+class AdaptiveQueryExecSuite
+    extends SparkQueryCompareTestSuite
+        with AdaptiveSparkPlanHelper {
+
+  private def runAdaptiveAndVerifyResult(
+      spark: SparkSession, query: String): (SparkPlan, SparkPlan) = {
+
+    val dfAdaptive = spark.sql(query)
+    val planBefore = dfAdaptive.queryExecution.executedPlan
+    assert(planBefore.toString.startsWith("AdaptiveSparkPlan isFinalPlan=false"))
+
+    dfAdaptive.collect()
+    val planAfter = dfAdaptive.queryExecution.executedPlan
+    assert(planAfter.toString.startsWith("AdaptiveSparkPlan isFinalPlan=true"))
+    val adaptivePlan = planAfter.asInstanceOf[AdaptiveSparkPlanExec].executedPlan
+
+    val exchanges = adaptivePlan.collect {
+      case e: Exchange => e
+    }
+    assert(exchanges.isEmpty, "The final plan should not contain any Exchange node.")
```

Comment:
Why should there be no exchanges?  That feels like a somewhat arbitrary requirement for a generically named function like `runAdaptiveAndVerifyResult`

Reply:
With AQE, the query is broken down into query stages based on exchange boundaries, so the final query that is executed depends on the results from its child query stages. There cannot be any exchange nodes left when the final query is executed because they will have already been replaced with QueryStageExecs.

Reply:
Then add that as a comment so it is clear what is happening and why this check is in place.